File: frontend/src/types/sale.ts
export type SaleStatus = 'pending' | 'confirmed' | 'processing' | 'completed' | 'cancelled';
export type PaymentMethod = 'cash' | 'transfer' | 'card' | 'qr' | 'mixed';
export type PaymentStatus = 'pending' | 'partial' | 'paid' | 'refunded' | 'failed';

export interface SaleItem {
  id: string;
  product_id: string;
  nombre_producto: string;
  sku_producto?: string;
  cantidad: number;
  precio_unitario: number;
  descuento: number;
  subtotal: number;
  iva_porcentaje: number;
  iva_monto: number;
  ice_porcentaje: number;
  ice_monto: number;
}

export interface SaleItemCreate {
  product_id: string;
  cantidad: number;
  precio_unitario?: number;
  descuento?: number;
}

export interface Sale {
  id: string;
  business_id: string;
  numero_venta: string;
  fecha_venta: string;
  estado: SaleStatus;
  subtotal: number;
  descuento: number;
  iva: number;
  ice: number;
  total: number;
  metodo_pago: PaymentMethod;
  estado_pago: PaymentStatus;
  fecha_pago?: string;
  factura_id?: string;
  pdf_url?: string;
  tipo_comprobante?: 'CONSUMIDOR_FINAL' | 'FACTURA';
  customer_id?: string;
  customer_name?: string;
  customer_phone?: string;
  customer_email?: string;
  customer_identification?: string;
  notas?: string;
  notas_internas?: string;
  items: SaleItem[];
  created_at: string;
  created_by?: string;
  confirmado_en?: string;
  completado_en?: string;
  cancelado_en?: string;
  motivo_cancelacion?: string;
}

export interface SaleCreate {
  customer_id?: string;
  customer_phone?: string;
  customer_email?: string;
  tipo_comprobante?: 'CONSUMIDOR_FINAL' | 'FACTURA';
  items: SaleItemCreate[];
  metodo_pago?: PaymentMethod;
  descuento?: number;
  tipo_descuento?: 'percentage' | 'fixed';
  notas?: string;
  enviar_whatsapp?: boolean;
}

export interface SaleUpdate {
  estado?: SaleStatus;
  metodo_pago?: PaymentMethod;
  estado_pago?: PaymentStatus;
  notas?: string;
  notas_internas?: string;
  motivo_cancelacion?: string;
}

export interface SaleListResponse {
  id: string;
  numero_venta: string;
  fecha_venta: string;
  customer_name?: string;
  customer_phone?: string;
  estado: SaleStatus;
  total: number;
  estado_pago: PaymentStatus;
  factura_id?: string;
}

File: backend/app/schemas/sale.py
"""
Sale schemas
"""
from pydantic import BaseModel, Field, ConfigDict, validator
from typing import Optional, List, Dict, Any
from uuid import UUID
from datetime import datetime
from decimal import Decimal

from app.constants.sales import SaleStatus, PaymentMethod, PaymentStatus


class SaleItemBase(BaseModel):
    product_id: UUID
    cantidad: int = Field(..., gt=0)
    precio_unitario: Optional[Decimal] = Field(None, gt=0)
    descuento: Decimal = Field(0, ge=0, decimal_places=2)


class SaleItemCreate(SaleItemBase):
    pass


class SaleItemResponse(BaseModel):
    id: UUID
    nombre_producto: str
    sku_producto: Optional[str]
    cantidad: int
    precio_unitario: Decimal
    descuento: Decimal
    subtotal: Decimal
    iva_porcentaje: Decimal
    iva_monto: Decimal
    ice_porcentaje: Decimal
    ice_monto: Decimal
    
    model_config = ConfigDict(from_attributes=True)


class SaleBase(BaseModel):
    customer_id: Optional[UUID] = None
    customer_email: Optional[str] = None
    metodo_pago: PaymentMethod = PaymentMethod.CASH
    tipo_comprobante: Optional[str] = None  # CONSUMIDOR_FINAL o FACTURA
    notas: Optional[str] = None
    notas_internas: Optional[str] = None


class SaleCreate(SaleBase):
    items: List[SaleItemCreate]
    descuento: Decimal = Field(0, ge=0, decimal_places=2)
    tipo_descuento: Optional[str] = None
    send_whatsapp: bool = True


class SaleUpdate(BaseModel):
    estado: Optional[SaleStatus] = None
    metodo_pago: Optional[PaymentMethod] = None
    estado_pago: Optional[PaymentStatus] = None
    notas: Optional[str] = None
    notas_internas: Optional[str] = None
    motivo_cancelacion: Optional[str] = None


class SaleResponse(SaleBase):
    id: UUID
    business_id: UUID
    numero_venta: str
    fecha_venta: datetime
    estado: SaleStatus
    subtotal: Decimal
    descuento: Decimal
    iva: Decimal
    ice: Decimal
    total: Decimal
    estado_pago: PaymentStatus
    factura_id: Optional[UUID]
    pdf_url: Optional[str] = None
    
    items: List[SaleItemResponse] = []
    
    # Customer info - usando property para acceder a la relación
    customer_name: Optional[str] = None
    customer_phone: Optional[str] = None
    customer_identification: Optional[str] = None
    
    created_at: datetime
    created_by: Optional[UUID]
    confirmado_en: Optional[datetime]
    completado_en: Optional[datetime]
    cancelado_en: Optional[datetime]
    
    model_config = ConfigDict(from_attributes=True)

    @classmethod
    def from_orm_with_customer(cls, sale):
        """Método factory para crear response con datos del customer"""
        data = sale.__dict__.copy()
        # Incluir pdf_url desde factura si existe
        if hasattr(sale, 'factura') and sale.factura:
            data['pdf_url'] = sale.factura.pdf_url if hasattr(sale.factura, 'pdf_url') else None
        if sale.customer:
            data['customer_name'] = sale.customer.name
            data['customer_phone'] = sale.customer.phone_number
            data['customer_email'] = sale.customer.email
            data['customer_identification'] = sale.customer.identification
        return cls(**data)
        

class SaleListResponse(BaseModel):
    id: UUID
    numero_venta: str
    fecha_venta: datetime
    customer_name: Optional[str]
    customer_phone: Optional[str]
    estado: SaleStatus
    total: Decimal
    estado_pago: PaymentStatus
    factura_id: Optional[UUID]
    
    model_config = ConfigDict(from_attributes=True)


class SaleStatusUpdate(BaseModel):
    estado: SaleStatus
    motivo_cancelacion: Optional[str] = Field(None, min_length=3)
    
    @validator('motivo_cancelacion')
    def validate_cancellation_reason(cls, v, values):
        if values.get('estado') == SaleStatus.CANCELLED and not v:
            raise ValueError('Cancellation reason is required when cancelling a sale')
        return v